### PR TITLE
Assignment count mismatch inspection

### LINKED ISSUE #1219: Assignment count mismatch should be reported
```go
package main
var c, d = func() { }
```

### LINKED ISSUE #1842: Multiple assignment error is not reported
```go
func Foo() (int, int) {
	a, b := 0, 0
	a, b = 1, 1
	a, b += 1, 1 // Should be error: `syntax error: unexpected op=, expecting := or = or comma`
	return a, b
}
```

### LINKED ISSUE #1906: Improve multiple-value in single-value context detection
```go
package main

func doubleReturn() (string, string) {
  return "foo", "bar";
}

func main() {
  _ = doubleReturn() // should be an error

  var existingVar string
  existingVar = doubleReturn() // should be an error

  println(existingVar)
}
```

### PR DESCRIPTION
**Pull request by dlsniper**
fixes #1219
fixes #1842
fixes #1906

### CODE REVIEW

**Comment by dlsniper** [comment]
I've yet to run the tests locally so I'll the CI do this for me.

**Comment by zolotov** [comment]
@dlsniper what's wrong with your local running? There is no difference between local and CI running, both uses gradlew-script

**Comment by dlsniper** [comment]
@zolotov I'm running on IDEA 16 EAP and for some reason (Kotlin differences) nothing that's based against master (IDEA 15) works, either when running `./gradlew runIdea` or `./gradlew cleanTest test`. I could jump to IDEA 15 and do it but it's a bit more work to jump versions (I'll do it if something fails on the CI run, ofc).

**Comment by zolotov** [comment]
Running tests and `gradlew` script have nothing to do with the IDEA you're running on, try to clean gradles cache (`~/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/`) in case you replaced kotlin runtime libraries.

**Comment by dlsniper** [comment]
That worked, thank you. I can run tests from the command line (and IDE) now.

File: src/com/goide/psi/impl/GoPsiImplUtil.java [hunk 1]
```diff
@@ -568,6 +566,36 @@ private static GoType findTypeInVarSpec(@NotNull GoVarDefinition o, @Nullable Re
...
+    // because we've essentially bumped into a function signature that's returned by a function
+    // e.g. func demo() func() (res1 int, res2 error)
+
+    if (type == null) return false;
+
+    PsiElement parentFunc = type;
+
+    if (type.getParent() instanceof GoParameterDeclaration &&
+        type.getParent().getParent() instanceof GoParameters) {
+      parentFunc = type.getParent().getParent().getParent();
```
Comment (ignatov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: Are you kidding? :) `type.getParent().getParent().getParent()`

File: src/com/goide/psi/impl/GoPsiImplUtil.java [hunk 2]
```diff
@@ -1265,4 +1293,415 @@ public PsiElement getType(@NotNull GoTypeSpec o) {
...
+    if (parent instanceof GoReferenceExpression || parent instanceof GoSelectorExpr) return false;
+    // A situation like T("foo").
+    return !(parent instanceof GoCallExpr);
+  }
+
+  public static boolean hasUnresolvedReferences(@NotNull List<GoExpression> expressionList) {
+    for (GoExpression expression : expressionList) {
+      if (expression instanceof GoReferenceExpression) {
+        if (expression.getReference() == null ||
+            expression.getReference().resolve() == null) {
```
Comment (ignatov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: Please extract a variable and simplify.

File: src/com/goide/psi/impl/GoPsiImplUtil.java [hunk 3]
```diff
@@ -1265,4 +1293,415 @@ public PsiElement getType(@NotNull GoTypeSpec o) {
...
+    return false;
+  }
+
+  public static boolean hasUnresolvedCalls(@NotNull List<GoExpression> expressionList) {
+    for (GoExpression expression : expressionList) {
+      if (expression instanceof GoCallExpr) {
+        expression = ((GoCallExpr)expression).getExpression();
+        if (expression.getGoType(null) == null) {
+          if (expression.getReference() != null) {
+            if (expression.getReference().resolve() == null) {
```
Comment (ignatov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: Please simplify

File: tests/com/goide/inspections/AssignmentCountMismatchTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.goide.inspections;
+
+import com.goide.GoCodeInsightFixtureTestCase;
+import com.intellij.testFramework.LightProjectDescriptor;
+
+public class AssignmentCountMismatchTest extends GoCodeInsightFixtureTestCase {
```
Comment (ignatov) [inline tests/com/goide/inspections/AssignmentCountMismatchTest.java]: Should be GoAssignmentCountMismatchTest

**Comment by dlsniper** [comment]
Hi @ignatov, sorry for delay, I've updated the PR per review feedback. Thank you

**Comment by dlsniper** [comment]
as this is beyond huge / ugly, I'll close it for now to declutter the list of PRs.

**Pull request closed without merging**